library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,425 @@
+    # - `#cleanup` [nil (String)] cleanup after widget is destroyed
+    # @raise [RuntimeError] if required method is not implemented or widget id not set.
+    def description
+      if widget_id.nil?
+        raise "Widget '#{self.class}' does set its widget ID"
+      end
+
+      res = {}
+
+      if respond_to?(:help)
+        res["help"] = help
+      else
+        res["no_help"] = ""
+      end
+      res["label"] = label if respond_to?(:label)
+      res["opt"] = opt if respond_to?(:opt)
+      if respond_to?(:validate)
+        res["validate_function"] = validate_method
+        res["validate_type"] = :function
+      end
+      res["init"] = init_method if respond_to?(:init)
+      res["handle"] = handle_method if respond_to?(:handle)
+      res["store"] = store_method if respond_to?(:store)
+      res["cleanup"] = cleanup_method if respond_to?(:cleanup)
+
+      res
+    end
+
+  protected
+    # shortcut from Yast namespace to avoid including whole namespace
+    # TODO: kill converts in CWM module, to avoid this workaround for funrefs
+    def fun_ref(*args)
+      Yast::FunRef.new(*args)
+    end
+
+  private
+
+    def init_method
+      fun_ref(method(:init), "void (string)")
+    end
+
+
+    def handle_method
+      fun_ref(method(:handle), "symbol (string, map)")
+    end
+
+    def store_method
+      fun_ref(method(:store), "void (string, map)")
+    end
+
+    def cleanup_method
+      fun_ref(method(:cleanup), "void (string)")
+    end
+
+    def validate_method
+      fun_ref(method(:validate), "boolean (string, map)")
+    end
+  end
+
+  class CustomWidget < AbstractWidget
+    def description
+      unless respond_to?(:content)
+        raise "For custom widget '#{self.class}' content method have to be defined"
+      end
+
+      {
+        "custom_widget" => content,
+        "widget"        => :custom
+      }.merge(super)
+    end
+  end
+
+  class EmptyWidget < AbstractWidget
+    def initialize(id)
+      self.widget_id = id
+    end
+
+    def description
+      {
+        "widget"        => :empty
+      }.merge(super)
+    end
+  end
+
+  # helpers for easier set/obtain value of widget for widgets where value is
+  # obtained by :Value symbol
+  module ValueBasedWidget
+
+  protected
+
+    def value
+      Yast::UI.QueryWidget(Id(widget_id), :Value)
+    end
+
+    def value=(val)
+      Yast::UI.ChangeWidget(Id(widget_id), :Value, val)
+    end
+  end
+
+  class InputFieldWidget < AbstractWidget
+    include ValueBasedWidget
+
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
+
+      {
+        "widget"        => :inputfield
+      }.merge(super)
+    end
+  end
+
+  class PasswordWidget < AbstractWidget
+    include ValueBasedWidget
+
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
```

Comment:
well, on other hand abstract method is third party gem, right? and we want to have minimal dependencies in yast2.